Xml2xml

epifanio opened the pull request:
This PR does a refactoring of the code to work with XML files. The code is as general as possible so that some of the functionalities can be applied not just to MMD and/or ISO but to any XML (hence the renaming of methods and files)
This P.R. also includes a new script to ruin transformation in parallel and adds few more tests, so to reach full coverage.

```nosetests --with-coverage --cover-package=py_mmd_tools```

```
................
Name                       Stmts   Miss  Cover   Missing
--------------------------------------------------------
py_mmd_tools/__init__.py       0      0   100%
py_mmd_tools/log_util.py      31      0   100%
py_mmd_tools/xml2xml.py       28      0   100%
py_mmd_tools/xml_util.py      24      0   100%
--------------------------------------------------------
TOTAL                         83      0   100%
----------------------------------------------------------------------
Ran 16 tests in 30.550s

OK
```
other minor changes in this P.R.:
* It is using the new data directory to load the test dataset and removes the dependency from the `confuse` 
* the logging utility methods accept logging to file or as std output - to avoid issue when running in parallel.

mortenwh commented on py_mmd_tools/xml2xml.py:
```diff
@@ -0,0 +1,57 @@
+import errno
+import os
+
+def xml2xml(
```
`xml_translate`

mortenwh commented on py_mmd_tools/xml2xml.py:
```diff
@@ -0,0 +1,57 @@
+    xsd_validation=False,
+    xsd_schema=None,
+):
+    """[Transform MMD file to ISO using xslt]
```
needs to be updated

mortenwh commented on py_mmd_tools/log_util.py:
```diff
@@ -0,0 +1,46 @@
+"""
+Tool for converting metadata from MMD format to ISO format using a specific xslt.
+ License:
+     This file is part of the S-ENDA-Prototype repository (https://github.com/metno/py-mmd-tools).
```
Should be: `License: This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)`

mortenwh commented on py_mmd_tools/log_util.py:
```diff
@@ -0,0 +1,46 @@
+import os
+
+def get_logpath(logdirpath):
+    if not pathlib.Path(logdirpath).exists():
```
Please add a docstring

mortenwh commented on py_mmd_tools/log_util.py:
```diff
@@ -0,0 +1,46 @@
+            raise IOError
+
+
+def setup_log(name, logdirpath, logtype='stream'):
```
please add docstring

mortenwh commented on py_mmd_tools/log_util.py:
```diff
@@ -0,0 +1,46 @@
+"""
+Tool for converting metadata from MMD format to ISO format using a specific xslt.
```
Seems this file is for something else?

mortenwh commented on py_mmd_tools/log_util.py:
```diff
@@ -0,0 +1,46 @@
+        ch.setLevel(logging.INFO)
+        ch.setFormatter(log_format)
+        logger.addHandler(ch)
+        return logger
```
Repetition - can you unindent `return logger`, and delete line 39?

mortenwh commented on py_mmd_tools/xml2xml.py:
```diff
@@ -0,0 +1,57 @@
+"""
+Tool for converting metadata from MMD format to ISO format using a specific xslt.
```
This is now a generic tool, isn't it? Can you update the docstring?

mortenwh commented on py_mmd_tools/xml2xml.py:
```diff
@@ -0,0 +1,57 @@
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), xsd_schema)
+        else: 
+            if not xsd_check(xml_file, xsd_schema=xsd_schema):
+                return False
```
you have three lines that return either True or False. Wouldn't it be better to assign the result to a variable, and return it in the end of the method?

mortenwh commented on py_mmd_tools/xml_util.py:
```diff
@@ -0,0 +1,58 @@
+"""
+Utility tool to work on mmd xml files.
+ License:
+     This file is part of the S-ENDA-Prototype repository (https://github.com/metno/py-mmd-tools).
```
Please change to License: `This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)`

mortenwh commented on py_mmd_tools/xml_util.py:
```diff
@@ -0,0 +1,58 @@
+    """
+    if pathlib.Path(xml_file).is_file() and os.path.getsize(xml_file) != 0:
+        try:
+            xml = ET.parse(xml_file)
```
isn't it enough with this line? FileNotFound and XMLSyntaxError should be raised by `ET.parse`, shouldn't it?

epifanio replied:
I may be wrong, but it looks like `ET.parse` raises a `OSerror`  when a non existent file is given directly, example below.:

```python
In [1]: import lxml.etree as ET  
In [2]: xml = ET.parse('tests/data/this_fine_not_exist.xml')                                                                                                                  
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-7-60df23105c82> in <module>
----> 1 xml = ET.parse('tests/data/this_fine_not_exist.xml')

src/lxml/etree.pyx in lxml.etree.parse()

src/lxml/parser.pxi in lxml.etree._parseDocument()

src/lxml/parser.pxi in lxml.etree._parseDocumentFromURL()

src/lxml/parser.pxi in lxml.etree._parseDocFromFile()

src/lxml/parser.pxi in lxml.etree._BaseParser._parseDocFromFile()

src/lxml/parser.pxi in lxml.etree._ParserContext._handleParseResultDoc()

src/lxml/parser.pxi in lxml.etree._handleParseResult()

src/lxml/parser.pxi in lxml.etree._raiseParseError()

OSError: Error reading file 'tests/data/not_a_valid_xmlh.xml': failed to load external entity "tests/data/this_fine_not_exist.xml"
```

mortenwh commented on py_mmd_tools/xml_util.py:
```diff
@@ -0,0 +1,58 @@
+        [bool]: [return True if a valid xml filepath is provided, 
+        return False if the xmlfile is invalid, empty, or doesn't exist ]
+    """
+    if not pathlib.Path(xsd_schema).is_file() and os.path.getsize(xsd_schema) != 0:
```
Seems this and the next three lines are obsolete? Don't lines 53-55 raise the correct exceptions?

mortenwh commented on py_mmd_tools/xml_util.py:
```diff
@@ -0,0 +1,58 @@
+            xmlschema_mmd = ET.XMLSchema(ET.parse(xsd_schema))
+            xml_doc = ET.ElementTree(file=xml_file)
+            if not xmlschema_mmd.validate(xml_doc):
+                return False
```
I prefer one return statement, and a variable assignment..

mortenwh commented on script/xmlconverter.py:
```diff
@@ -0,0 +1,141 @@
+Script to run the xmlc onversion tools
+
+ License:
+     This file is part of the S-ENDA-Prototype repository (https://github.com/metno/py-mmd-tools).
```
please change to `License: This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)`

mortenwh commented on script/xmlconverter.py:
```diff
@@ -0,0 +1,141 @@
+                xml_files.append(file_path)
+    return xml_files
+
+def writerecord(xml_file, xslt, outdir="/tmp"):
```
perhaps `translate_and_write`? (it does more than just writing the record..)

mortenwh commented on script/xmlconverter.py:
```diff
@@ -0,0 +1,141 @@
+        for i in xmlfiles:
+            try:
+                writerecord(xml_file=i, xslt=xslt, outdir=outdir)
+            except XMLSyntaxError:
```
adding `as e` in this line, and `e.message` in the next might be useful

mortenwh commented on script/xmlconverter.py:
```diff
@@ -0,0 +1,141 @@
+                print(f"failed on: {i}")
+
+
+def check_record(record, tobedone):
```
please add docstring

mortenwh commented on script/xmlconverter.py:
```diff
@@ -0,0 +1,141 @@
+        return record
+
+
+def recover_task(metadata, outdir, parallel=False):
```
docstring needed

mortenwh commented on tests/test_xml_util.py:
```diff
@@ -0,0 +1,49 @@
+        self.not_a_valid_xml = str(current_dir / 'tests' / 'data' / 'not_a_valid_xml.xml')
+ 
+    def test_xml_check_1(self):
+        unittest.TestCase.assertEqual(self, xml_check(xml_file=self.reference_xml), 
```
`self.assertEqual(xml_check(xml_file=self.reference_xml), True)` -  isn't this working?

Or even better: `self.assertTrue(xml_check(xml_file=self.reference_xml))`

mortenwh reviewed (commented):
Please check inline comments. There's a few changes that needs to be done.

epifanio commented:
I addressed most of your comments. Making a new PR with the changes - Note that the proposed changes affected code coverage a bit - I didn't manage to get it higher has it was before.
